app/helpers/application_helper.rb

```diff
@@ -322,20 +322,35 @@ def link_to_chart(name, values)
     link_to icon_tag('signal'), url, target: :blank
   end
 
-  # show which stages this reference is deploy(ed+ing) to
+  # Show which stages this reference has been or is currently being deployed to.
   def deployed_or_running_list(stages, reference)
-    html = "".html_safe
-    stages.each do |stage|
-      next unless deploy = stage.deployed_or_running_deploy
-      next unless deploy.references?(reference)
-      label = (deploy.active? ? "label-warning" : "label-success")
-
-      text = "".html_safe
-      text << stage.name
-      html << content_tag(:span, text, class: "label #{label} release-stage")
-      html << " "
+    deploys = Deploy.of_reference_in_stages(reference, stages)
```

Comment:
```Ruby
deploy_lookup = Deploy.where(reference: reference).where(stage_id: stages.map(&:id)).group(:stage_id).      select("MAX(id) AS id, stage_id").group_by(&:stage_id)
```

Reply:
I really dislike putting SQL in the view layer – it makes testing so much harder. I'd very much prefer to keep this in the model.

Reply:
the query logic is closely tied to the compare logic, so could lead to bugs/get out of sync when it's extracted to the model ... but I don't care that much
more important: the group_by line seems much easier to read (assuming it works as before) ?